benchclients/python/benchclients/http.py

```diff
@@ -0,0 +1,414 @@
+import logging
+import time
+from abc import ABC, abstractmethod
+from json import dumps as jsondumps
+from typing import Dict, List, Literal, Optional, Tuple, Union
+
+import requests
+
+log = logging.getLogger(__name__)
+
+
+TypeHTTPMethods = Literal["GET", "PUT", "POST", "DELETE", "HEAD", "OPTIONS", "HEAD"]
+
+
+class RetryingHTTPClientException(Exception):
+    """
+    Base exception type used by RetryingHTTPClient.
+    """
+
+
+class RetryingHTTPClientLoginError(RetryingHTTPClientException):
+    """
+    Base exception for login errors, raise directly when login failed as of any
+    non-401 error (network problem, other unexpected response, ...).
+    """
+
+
+class RetryingHTTPClientBadCredentials(RetryingHTTPClientLoginError):
+    """
+    Raise when login fails as of a 401 error (bad credentials).
+    """
+
+
+class RetryingHTTPClientDeadlineReached(RetryingHTTPClientException):
+    """
+    Raise when giving up after retrying as of a per-request retrying
+    deadline/timeout criterion, usually shortly before reaching the deadline
+    (rarely exceeding).
+    """
+
+
+class RetryingHTTPClientNonRetryableResponse(RetryingHTTPClientException):
+    """
+    Raise when within a retry loop and when receiving an HTTP response that
+    should not be repeated.
+    """
+
+    def __init__(self, message, error_response: requests.Response):
+        super().__init__(message)
+        self.error_response = error_response
+
+
+class RetryingHTTPClient(ABC):
+    """
+    HTTP client abstraction tuned towards use cases in the context of
+    continuous integration jobs.
+
+    - focus on persistent retrying
+    - detailed logging along the request/response retrying lifecycle
+    - handle 401 response with re-login
+    """
+
+    default_retry_for_seconds: float
+    timeout_login_request: Tuple[float, float]
+    timeout_long_running_requests: Tuple[float, float]
+
+    # timeout_connect_seconds = 3.5
+    # Note(JP): I have bumped this from 10 to 75 seconds to err on side of
+    # caution (remove stress from DB, at the cost of potentially longer-running
+    # jobs, and at the cost of time-between-useful-logmsgs). This needs more
+    # context-specific timeout constants, also see
+    # https://github.com/conbench/conbench/issues/801 and
+    # https://github.com/conbench/conbench/issues/806
+    # timeout_recv_seconds = 120
```

Comment:
nit: remove this comment block since you put it in ConbenchClient

Reply:
thanks!